# Pandas Practical Learning

About: **Pandas** is a Python library that provides extensive means for data analysis. Most of the time the data to be analyzed is stored in table formats like .csv, .tsv, or .xlsx. Pandas makes it 
- very convenient to load, process, and analyze such tabular data using SQL-like queries. 
- It has functions for analyzing, cleaning, exploring, and manipulating data.
- In conjunction with Matplotlib and Seaborn, Pandas provides a wide range of opportunities for visual analysis of tabular data.

Data structures in Pandas: 
- Series:  a one-dimensional indexed array of some fixed data type
- DataFrame classes: a two-dimensional data structure - a table - where each column contains data of the same type

In [25]:
import numpy as np
import pandas as pd

In [26]:
#Reading the data present in csv form using read_csv
df = pd.read_csv("telecom_churn.csv")

The first in data anaylsis is to get familiar with the structure of data i.e what different kind of data is available to us. Since most of the size of data set we are analyzing is very large only having a look at the first few entries is enough for us. To take a look at the first 5 entries we use **.head()** function.

In [27]:
df.head()

State  Account length  Area code International plan Voice mail plan  \
0    KS             128        415                 No             Yes   
1    OH             107        415                 No             Yes   
2    NJ             137        415                 No              No   
3    OH              84        408                Yes              No   
4    OK              75        415                Yes              No   

   Number vmail messages  Total day minutes  Total day calls  \
0                     25              265.1              110   
1                     26              161.6              123   
2                      0              243.4              114   
3                      0              299.4               71   
4                      0              166.7              113   

   Total day charge  Total eve minutes  Total eve calls  Total eve charge  \
0             45.07              197.4               99             16.78   
1             27.47              195.5              103             16.62   
2             41.38              121.2              110             10.30   
3             50.90               61.9               88              5.26   
4             28.34              148.3              122             12.61   

   Total night minutes  Total night calls  Total night charge  \
0                244.7                 91               11.01   
1                254.4                103               11.45   
2                162.6                104                7.32   
3                196.9                 89                8.86   
4                186.9                121                8.41   

   Total intl minutes  Total intl calls  Total intl charge  \
0                10.0                 3               2.70   
1                13.7                 3               3.70   
2                12.2                 5               3.29   
3                 6.6                 7               1.78   
4                10.1                 3               2.73   

   Customer service calls  Churn  
0                       1  False  
1                       1  False  
2                       0  False  
3                       2  False  
4                       3  False

Observations: Each row corresponds to one client or an instance, and columns are features of this instance.

Before performing any operations we must be aware of the shape of the data (i.e number of row & col). For this we use **.shape** function which returns a tuples with two values:

- first one represent number of rows (no. of examples)

- second one represent number of cols (no. of features)

In [28]:
print(df.shape)

(3333, 20)


Now let's try printing out column names using **.columns**.

In [29]:
print(df.columns)

Index(['State', 'Account length', 'Area code', 'International plan',
       'Voice mail plan', 'Number vmail messages', 'Total day minutes',
       'Total day calls', 'Total day charge', 'Total eve minutes',
       'Total eve calls', 'Total eve charge', 'Total night minutes',
       'Total night calls', 'Total night charge', 'Total intl minutes',
       'Total intl calls', 'Total intl charge', 'Customer service calls',
       'Churn'],
      dtype='object')


## Viewing general info
To get some general information about dataframe we can use **.info()**, which would help us to better understand the data. Like *datatypes* of the feature values or if there are any *null values* present in the data (i.e missing).

In [30]:
print(df.info())

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 3333 entries, 0 to 3332
Data columns (total 20 columns):
 #   Column                  Non-Null Count  Dtype  
---  ------                  --------------  -----  
 0   State                   3333 non-null   object 
 1   Account length          3333 non-null   int64  
 2   Area code               3333 non-null   int64  
 3   International plan      3333 non-null   object 
 4   Voice mail plan         3333 non-null   object 
 5   Number vmail messages   3333 non-null   int64  
 6   Total day minutes       3333 non-null   float64
 7   Total day calls         3333 non-null   int64  
 8   Total day charge        3333 non-null   float64
 9   Total eve minutes       3333 non-null   float64
 10  Total eve calls         3333 non-null   int64  
 11  Total eve charge        3333 non-null   float64
 12  Total night minutes     3333 non-null   float64
 13  Total night calls       3333 non-null   int64  
 14  Total night charge      3333 non-null   

### Observations:

- bool, int64, float64 and object are the data types of our features. 
- one feature is logical (bool), 3 features are of type object, and 16 features are numeric
- No missing values because each column contains 3333 observations, the same number of rows we saw before with shape.

## Getting the stats about the data:

Knowing some stats about the data always help in analyzation. To get some we can use the **describe()** function of the pandas. Let's see how that works:

In [31]:
df.describe()

Account length    Area code  Number vmail messages  Total day minutes  \
count     3333.000000  3333.000000            3333.000000        3333.000000   
mean       101.064806   437.182418               8.099010         179.775098   
std         39.822106    42.371290              13.688365          54.467389   
min          1.000000   408.000000               0.000000           0.000000   
25%         74.000000   408.000000               0.000000         143.700000   
50%        101.000000   415.000000               0.000000         179.400000   
75%        127.000000   510.000000              20.000000         216.400000   
max        243.000000   510.000000              51.000000         350.800000   

       Total day calls  Total day charge  Total eve minutes  Total eve calls  \
count      3333.000000       3333.000000        3333.000000      3333.000000   
mean        100.435644         30.562307         200.980348       100.114311   
std          20.069084          9.259435          50.713844        19.922625   
min           0.000000          0.000000           0.000000         0.000000   
25%          87.000000         24.430000         166.600000        87.000000   
50%         101.000000         30.500000         201.400000       100.000000   
75%         114.000000         36.790000         235.300000       114.000000   
max         165.000000         59.640000         363.700000       170.000000   

       Total eve charge  Total night minutes  Total night calls  \
count       3333.000000          3333.000000        3333.000000   
mean          17.083540           200.872037         100.107711   
std            4.310668            50.573847          19.568609   
min            0.000000            23.200000          33.000000   
25%           14.160000           167.000000          87.000000   
50%           17.120000           201.200000         100.000000   
75%           20.000000           235.300000         113.000000   
max           30.910000           395.000000         175.000000   

       Total night charge  Total intl minutes  Total intl calls  \
count         3333.000000         3333.000000       3333.000000   
mean             9.039325           10.237294          4.479448   
std              2.275873            2.791840          2.461214   
min              1.040000            0.000000          0.000000   
25%              7.520000            8.500000          3.000000   
50%              9.050000           10.300000          4.000000   
75%             10.590000           12.100000          6.000000   
max             17.770000           20.000000         20.000000   

       Total intl charge  Customer service calls  
count        3333.000000             3333.000000  
mean            2.764581                1.562856  
std             0.753773                1.315491  
min             0.000000                0.000000  
25%             2.300000                1.000000  
50%             2.780000                1.000000  
75%             3.270000                2.000000  
max             5.400000                9.000000

### Observations:

The describe method shows basic statistical characteristics of each numerical feature (int64 and float64 types): number of non-missing values, mean, standard deviation, range, median, 0.25 and 0.75 quartiles.

**NOTE:** One thing to notice is that our describe function didn't give any statistical observation on **Churn** feature. The reason is being that, the describe function by default **.describe()** only works for numerical values as most of the stat values can't be calculated for a bool values.

In order to see statistics on non-numerical features(such as **bool values**), one has to explicitly indicate data types of interested feature in the include parameter.

In [32]:
df.describe(include=["object", "bool"])

State International plan Voice mail plan  Churn
count   3333               3333            3333   3333
unique    51                  2               2      2
top       WV                 No              No  False
freq     106               3010            2411   2850

## Observations:

count - number of examples available of the feature

unique - number of unique values of the feature

top - the most ocurring value 

freq - frequency of the most occuring value

To see the count of a each value in the feature, we can use .value_counts().

Let's have a look at the distribution of Churn:

In [33]:
df["Churn"].value_counts()

False    2850
True      483
Name: Churn, dtype: int64

## Performing various operations on the pandas dataFrames

#### Sorting:

A DataFrame can be sorted by the value of one of the variables/features (i.e columns). For example, we can sort by Total day charge (use ascending=False to sort in descending order):

In [34]:
result = df.sort_values(by="Total day charge", ascending=False).head()

We can also sort by multiple columns:

In [35]:
#outputs the data in the increasing order of Churn values but decreasing order of total day mins
# that means the first row would be the one which has the most daily minutes and least value of churn
df.sort_values(by=["Churn", "Total day minutes"], ascending=[True, False]).head()

State  Account length  Area code International plan Voice mail plan  \
688     MN              13        510                 No             Yes   
2259    NC             210        415                 No             Yes   
534     LA              67        510                 No              No   
575     SD             114        415                 No             Yes   
2858    AL             141        510                 No             Yes   

      Number vmail messages  Total day minutes  Total day calls  \
688                      21              315.6              105   
2259                     31              313.8               87   
534                       0              310.4               97   
575                      36              309.9               90   
2858                     28              308.0              123   

      Total day charge  Total eve minutes  Total eve calls  Total eve charge  \
688              53.65              208.9               71             17.76   
2259             53.35              147.7              103             12.55   
534              52.77               66.5              123              5.65   
575              52.68              200.3               89             17.03   
2858             52.36              247.8              128             21.06   

      Total night minutes  Total night calls  Total night charge  \
688                 260.1                123               11.70   
2259                192.7                 97                8.67   
534                 246.5                 99               11.09   
575                 183.5                105                8.26   
2858                152.9                103                6.88   

      Total intl minutes  Total intl calls  Total intl charge  \
688                 12.1                 3               3.27   
2259                10.1                 7               2.73   
534                  9.2                10               2.48   
575                 14.2                 2               3.83   
2858                 7.4                 3               2.00   

      Customer service calls  Churn  
688                        3  False  
2259                       3  False  
534                        4  False  
575                        1  False  
2858                       1  False

## Indexing and retrieving data :

When we want to work with the specific part of the data, retrieval of data becomes so crucial. Pandas provides efficient ways to retrieve tabular data. The most commnly used funcitons are:

### DataFrame['Name'] construction - To get all the values in a single column

For eg;

   1. Let's say we want to know which all unique states are present in the dataset.
    
   **OR**
    
   2. We want to calculate the average day minutes

In [131]:
# unique states
len(df['State'].unique())

51

In [129]:
#mean 
df['Total day minutes'].mean()

179.77509750975116

### Boolean indexing with one column - To get data/stats based on some condition
 
 For eg;
     
   1. Let's say we want to know the number of churned users with internation plan.
   2. What is the maximum length of international calls among loyal users (Churn == False) who do not have an international plan?

In [54]:
df[(df["Churn"] == True) & (df["International plan"] == "Yes")]["International plan"].count()

137

In [56]:
df[df["Churn"] == False]["Total intl minutes"].max()

18.9

### Indexing using loc and iloc method

DataFrames can be indexed by column name (label) or row name (index) or by the serial number of a row. The loc method is used for indexing by name, while iloc() is used for indexing by number.

1. Output values of the rows with index from 0 to 5 (inclusive) and columns labeled from State to Area code (inclusive)".

In [57]:
df.loc[0:5, "State":"Area code"]

State  Account length  Area code
0    KS             128        415
1    OH             107        415
2    NJ             137        415
3    OH              84        408
4    OK              75        415
5    AL             118        510

2. Give us the values of the first five rows in the first three columns

In [58]:
df.iloc[0:5, 0:3]

State  Account length  Area code
0    KS             128        415
1    OH             107        415
2    NJ             137        415
3    OH              84        408
4    OK              75        415

## Applying Functions to Cells, Columns and Rows

**To apply a function to each column we use apply()** - 
    
    for eg; To get the max value of each feature (i.e column) we can do the following

In [62]:
"""Here we have used np because we have to find the maximum value from a list of values for a particular feature"""
df.apply(np.max)

State                        WY
Account length              243
Area code                   510
International plan          Yes
Voice mail plan             Yes
Number vmail messages        51
Total day minutes         350.8
Total day calls             165
Total day charge          59.64
Total eve minutes         363.7
Total eve calls             170
Total eve charge          30.91
Total night minutes       395.0
Total night calls           175
Total night charge        17.77
Total intl minutes         20.0
Total intl calls             20
Total intl charge           5.4
Customer service calls        9
Churn                      True
dtype: object

#### To apply functions in rows (across horizontally) we have to use lambda functions.

For eg; We want the data of only those states whose name starts with N

In [132]:
df[df["State"].apply(lambda x: x[0] == "N")].head()

State  Account length  Area code International plan Voice mail plan  \
2     NJ             137        415                 No           False   
15    NY             161        415                 No           False   
25    NE             174        415                 No           False   
31    NH              75        510                 No           False   
39    NJ             149        408                 No           False   

    Number vmail messages  Total day minutes  Total day calls  \
2                       0              243.4              114   
15                      0              332.9               67   
25                      0              124.3               76   
31                      0              226.1              105   
39                      0              140.4               94   

    Total day charge  Total eve minutes  ...  Total eve charge  \
2              41.38              121.2  ...             10.30   
15             56.59              317.8  ...             27.01   
25             21.13              277.1  ...             23.55   
31             38.44              201.5  ...             17.13   
39             23.87              271.8  ...             23.10   

    Total night minutes  Total night calls  Total night charge  \
2                 162.6                104                7.32   
15                160.6                128                7.23   
25                250.7                115               11.28   
31                246.2                 98               11.08   
39                188.3                108                8.47   

    Total intl minutes  Total intl calls  Total intl charge  \
2                 12.2                 5               3.29   
15                 5.4                 9               1.46   
25                15.5                 5               4.19   
31                10.3                 5               2.78   
39                11.1                 9               3.00   

    Customer service calls  Churn  State_id  
2                        0  False        32  
15                       4   True        35  
25                       3  False        30  
31                       1  False        31  
39                       1  False        32  

[5 rows x 21 columns]

## Map method
The map method can be used to replace values in a column or creating a different column by mapping each unique values to a different value in by passing a dictionary of the form {old_value: new_value} as its argument:

Let's say we want to assign each state in alphabaetical order, a unique id starting from 1. We can follow the below steps:

In [75]:
unique_states = df["State"].unique()

In [76]:
unique_states.sort()
state_id = {}
i = 1
for state in unique_states:
    state_id[state] = i
    i += 1
print(state_id)

{'AK': 1, 'AL': 2, 'AR': 3, 'AZ': 4, 'CA': 5, 'CO': 6, 'CT': 7, 'DC': 8, 'DE': 9, 'FL': 10, 'GA': 11, 'HI': 12, 'IA': 13, 'ID': 14, 'IL': 15, 'IN': 16, 'KS': 17, 'KY': 18, 'LA': 19, 'MA': 20, 'MD': 21, 'ME': 22, 'MI': 23, 'MN': 24, 'MO': 25, 'MS': 26, 'MT': 27, 'NC': 28, 'ND': 29, 'NE': 30, 'NH': 31, 'NJ': 32, 'NM': 33, 'NV': 34, 'NY': 35, 'OH': 36, 'OK': 37, 'OR': 38, 'PA': 39, 'RI': 40, 'SC': 41, 'SD': 42, 'TN': 43, 'TX': 44, 'UT': 45, 'VA': 46, 'VT': 47, 'WA': 48, 'WI': 49, 'WV': 50, 'WY': 51}


In [77]:
df["State_id"] = df["State"].map(state_id)

In [78]:
df.head()

State  Account length  Area code International plan Voice mail plan  \
0    KS             128        415                 No             Yes   
1    OH             107        415                 No             Yes   
2    NJ             137        415                 No              No   
3    OH              84        408                Yes              No   
4    OK              75        415                Yes              No   

   Number vmail messages  Total day minutes  Total day calls  \
0                     25              265.1              110   
1                     26              161.6              123   
2                      0              243.4              114   
3                      0              299.4               71   
4                      0              166.7              113   

   Total day charge  Total eve minutes  ...  Total eve charge  \
0             45.07              197.4  ...             16.78   
1             27.47              195.5  ...             16.62   
2             41.38              121.2  ...             10.30   
3             50.90               61.9  ...              5.26   
4             28.34              148.3  ...             12.61   

   Total night minutes  Total night calls  Total night charge  \
0                244.7                 91               11.01   
1                254.4                103               11.45   
2                162.6                104                7.32   
3                196.9                 89                8.86   
4                186.9                121                8.41   

   Total intl minutes  Total intl calls  Total intl charge  \
0                10.0                 3               2.70   
1                13.7                 3               3.70   
2                12.2                 5               3.29   
3                 6.6                 7               1.78   
4                10.1                 3               2.73   

   Customer service calls  Churn  State_id  
0                       1  False        17  
1                       1  False        36  
2                       0  False        32  
3                       2  False        36  
4                       3  False        37  

[5 rows x 21 columns]

## replace Method:

The same thing can be done with the replace method. For eg; if we want to replace each **'yes'** with **'true'** and '**no**' with **'false'**, the we can do something like below

In [79]:
d = {"Yes": True, "No": "False"}
df = df.replace({"Voice mail plan": d})
df.head()

State  Account length  Area code International plan Voice mail plan  \
0    KS             128        415                 No            True   
1    OH             107        415                 No            True   
2    NJ             137        415                 No           False   
3    OH              84        408                Yes           False   
4    OK              75        415                Yes           False   

   Number vmail messages  Total day minutes  Total day calls  \
0                     25              265.1              110   
1                     26              161.6              123   
2                      0              243.4              114   
3                      0              299.4               71   
4                      0              166.7              113   

   Total day charge  Total eve minutes  ...  Total eve charge  \
0             45.07              197.4  ...             16.78   
1             27.47              195.5  ...             16.62   
2             41.38              121.2  ...             10.30   
3             50.90               61.9  ...              5.26   
4             28.34              148.3  ...             12.61   

   Total night minutes  Total night calls  Total night charge  \
0                244.7                 91               11.01   
1                254.4                103               11.45   
2                162.6                104                7.32   
3                196.9                 89                8.86   
4                186.9                121                8.41   

   Total intl minutes  Total intl calls  Total intl charge  \
0                10.0                 3               2.70   
1                13.7                 3               3.70   
2                12.2                 5               3.29   
3                 6.6                 7               1.78   
4                10.1                 3               2.73   

   Customer service calls  Churn  State_id  
0                       1  False        17  
1                       1  False        36  
2                       0  False        32  
3                       2  False        36  
4                       3  False        37  

[5 rows x 21 columns]

## Grouping

Groups the rows consisting same values for one or several features and return a DataFrameGroupby objects on which we can apply operations.

In general, grouping data in Pandas works as follows:

`df.groupby(by=grouping_columns)[columns_to_show].function()`

1. First, the groupby method divides the grouping_columns by their values. They become a new index in the resulting dataframe.
2. Then, columns of interest are selected (columns_to_show). If columns_to_show is not included, all non groupby clauses will be included.
3. Finally, one or several functions are applied to the obtained groups per selected columns.

In [85]:
columns_to_show = ["Total day minutes", "Total eve minutes", "Total night minutes"]

df.groupby(["Churn"])[columns_to_show].describe(percentiles=[])

Total day minutes                                            \
                  count        mean        std  min    50%    max   
Churn                                                               
False            2850.0  175.175754  50.181655  0.0  177.2  315.6   
True              483.0  206.914079  68.997792  0.0  217.6  350.8   

      Total eve minutes                                             \
                  count        mean        std   min    50%    max   
Churn                                                                
False            2850.0  199.043298  50.292175   0.0  199.6  361.8   
True              483.0  212.410145  51.728910  70.9  211.3  363.7   

      Total night minutes                                              
                    count        mean        std   min     50%    max  
Churn                                                                  
False              2850.0  200.133193  51.105032  23.2  200.25  395.0  
True                483.0  205.231677  47.132825  47.4  204.80  354.9

In [134]:
df.sort_values(by = ["State_id"]).groupby(["State"])[columns_to_show].describe(percentiles=[]).head()

Total day minutes                                              \
                  count        mean        std   min     50%    max   
State                                                                 
AK                 52.0  178.384615  49.640430  58.2  177.25  278.4   
AL                 80.0  186.010000  51.466249  68.7  190.25  308.0   
AR                 55.0  176.116364  50.368831  55.3  170.70  273.4   
AZ                 64.0  171.604688  51.941907  58.9  171.45  281.1   
CA                 34.0  183.564706  47.742484  92.8  183.20  280.0   

      Total eve minutes                                               \
                  count        mean        std    min     50%    max   
State                                                                  
AK                 52.0  184.282692  49.160213   58.6  179.95  314.4   
AL                 80.0  195.462500  50.909648   77.9  201.25  299.9   
AR                 55.0  201.047273  50.957484  120.5  192.30  350.9   
AZ                 64.0  187.748437  49.070513   72.9  191.10  328.7   
CA                 34.0  198.970588  40.361770  114.0  195.45  281.3   

      Total night minutes                                             
                    count        mean        std   min    50%    max  
State                                                                 
AK                   52.0  192.326923  53.663297  23.2  200.5  303.5  
AL                   80.0  187.285000  42.355416  94.1  185.4  287.6  
AR                   55.0  205.454545  56.696498  96.4  205.5  367.7  
AZ                   64.0  194.004687  54.387325  77.3  191.4  297.9  
CA                   34.0  198.508824  58.502574  71.1  204.9  345.8

In [121]:
ndf = df.sort_values("State_id").groupby(["State"])["Total day charge"].sum()

In [122]:
ndf

State
AK    1576.94
AL    2529.72
AR    1646.70
AZ    1867.08
CA    1061.01
CO    2005.20
CT    2203.30
DC    1573.31
DE    1810.49
FL    1922.85
GA    1698.58
HI    1585.39
IA    1328.60
ID    2216.70
IL    1711.67
IN    2372.11
KS    2279.56
KY    1742.81
LA    1546.56
MA    1990.14
MD    2347.07
ME    1952.74
MI    2241.17
MN    2618.36
MO    1826.20
MS    1966.15
MT    2011.55
NC    2140.32
ND    1974.61
NE    1840.33
NH    1688.19
NJ    2268.39
NM    1806.86
NV    1979.53
NY    2470.94
OH    2430.29
OK    1865.67
OR    2337.10
PA    1441.10
RI    1850.69
SC    1697.75
SD    1934.86
TN    1583.73
TX    2221.81
UT    2246.92
VA    2320.14
VT    2259.08
WA    2005.52
WI    2375.28
WV    3134.67
WY    2358.43
Name: Total day charge, dtype: float64

## Summary Tables using `crosstab`


Suppose we want to see how the observations in our sample are distributed in the context of two variables - `Churn` and `International plan`. To do so, we can build a contingency table using the `crosstab` method:

In [126]:
pd.crosstab(df["State"], df["Churn"]).head(10)

Churn  False  True
State             
AK        49     3
AL        72     8
AR        44    11
AZ        60     4
CA        25     9
CO        57     9
CT        62    12
DC        49     5
DE        52     9
FL        55     8